x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java

```diff
@@ -0,0 +1,293 @@
+                    if (leaderClusterState != null) {
+                        assert e == null;
+                        handleClusterAlias(clusterAlias, autoFollowPattern, followedIndices, leaderClusterState);
+                    } else {
+                        finalise(e);
+                    }
+                });
+            }
+        }
+
+        private void handleClusterAlias(String clusterAlias, AutoFollowPattern autoFollowPattern,
+                                List<String> followedIndexUUIDs, ClusterState leaderClusterState) {
+            List<Index> leaderIndicesToFollow = getLeaderIndicesToFollow(autoFollowPattern, leaderClusterState, followedIndexUUIDs);
+            if (leaderIndicesToFollow.isEmpty()) {
+                finalise(null);
+            } else {
+                CountDown countDown = new CountDown(leaderIndicesToFollow.size());
+                for (Index indexToFollow : leaderIndicesToFollow) {
+                    final String leaderIndexName = indexToFollow.getName();
+                    final String followIndexName = getFollowerIndexName(autoFollowPattern, leaderIndexName);
+
+                    String leaderIndexNameWithClusterAliasPrefix = clusterAlias.equals("_local_") ? leaderIndexName :
+                        clusterAlias + ":" + leaderIndexName;
+                    FollowIndexAction.Request followRequest =
+                        new FollowIndexAction.Request(leaderIndexNameWithClusterAliasPrefix, followIndexName,
+                            autoFollowPattern.getMaxBatchOperationCount(), autoFollowPattern.getMaxConcurrentReadBatches(),
+                            autoFollowPattern.getMaxOperationSizeInBytes(), autoFollowPattern.getMaxConcurrentWriteBatches(),
+                            autoFollowPattern.getMaxWriteBufferSize(), autoFollowPattern.getRetryTimeout(),
+                            autoFollowPattern.getIdleShardRetryDelay());
+
+                    // Execute if the create and follow api call succeeds:
+                    Runnable successHandler = () -> {
+                        LOGGER.info("Auto followed leader index [{}] as follow index [{}]", leaderIndexName, followIndexName);
+
+                        // This function updates the auto follow metadata in the cluster to record that the leader index has been followed:
+                        // (so that we do not try to follow it in subsequent auto follow runs)
+                        Function<ClusterState, ClusterState> function = recordLeaderIndexAsFollowFunction(clusterAlias, indexToFollow);
+                        // The coordinator always runs on the elected master node, so we can update cluster state here:
+                        updateAutoFollowMetadata(function, updateError -> {
+                            if (updateError != null) {
+                                LOGGER.error("Failed to mark leader index [" + leaderIndexName + "] as auto followed", updateError);
```

Comment:
So I am concerned about the error handling here. What happens if we have created the follower and actively started replication, but failed to add the index to the metadata? I think that it means we will retry on the next auto follower execution, which will then fail because the follower index already exists. And we will keep retrying. I think we need some protection here.

Reply:
> I think that it means we will retry on the next auto follower execution, which will then fail because the follower index already exists. And we will keep retrying.

Agreed. that would be a bad situation. I suspect such a thing is possible when the current master node fails.

Reply:
The only think that I can think of that we can do here without making this code more complicated if to have another background check that verifies that there is an entry for all leader indices being followed. We could check persistent tasks to figure this out or use the leader index metadata that we're going to add as custom to IndexMetaData. What do you think about this?

Reply:
Yeah, I am thinking something along those lines too. Like: if a leader index matches a pattern, and we are already following it, then we can skip create and follow and only need to get that index into the metadata. Your approach of using the follower index metadata about the leader index it is following sounds like a good approach to me.

Reply:
I think it then makes sense to do this in a follow up PR? (because the follower index metadata  needs to be added first)

Reply:
Yeah, I think so. I think we can put some skeleton code into place now, with a bunch of `TODOs`. Would you add this to the meta-issue?

Reply:
I added an entry to #33007